Added smartbanner code to suggest iphone/android apps for mobile users - fixes #9901

flhat opened the pull request:
Fixes #9901

### Changes
- Adds appropriate meta data to show smartbanner for iphone/android
- Includes iTunes / Google Play Store logo images in /static/icons/
- Adds /static/smartbanner/smartbanner.min.css and /static/smartbanner/smartbanner.min.js needed for the component

----
UUID: 495eed5c-2828-4e24-a19d-755ac5e95295

flhat committed:
Added smartbanner code to suggest iphone/android apps for mobile users

Alys commented:
@BlkStormy Thanks! Can you include a screenshot please?

flhat commented:
@Alys Sure - sorry, I forgot to do that with the message:

**iPhone**
<img width="372" alt="iphone" src="https://user-images.githubusercontent.com/4992601/44151817-85c48b38-a058-11e8-9ac7-96c83c412baa.png">

**Android**
<img width="371" alt="android" src="https://user-images.githubusercontent.com/4992601/44151818-85dcda12-a058-11e8-9d75-28fd6e607ef7.png">

flhat commented:
Hmm, I saw that after the pull request was made, it gets run through 2 different tests.  Both failed, but what I'm trying to figure out is if there's a way I can run the tests on my own locally, before doing the request?

TheHollidayInn commented:
Hey! Looks like your problem is with lint. You can run `npm run lint` but the issues is the new JS file you included. Is there a hosted version of the library on a CDN?

flhat commented:
@TheHollidayInn Hmm, I don't see one.  There is a NPM package here:  https://www.npmjs.com/package/smartbanner.js  I'm unfamiliar with how to get it to load once installed though...  Thoughts?

Alys commented:
@BlkStormy In case it's of use later, there's information about the tests in the "Creating and Running Automated Tests" section of [Using Your Local Install to Modify Habitica's Website and API](https://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API#Creating_and_Running_Automated_Tests)

TheHollidayInn commented:
@BlkStormy ah I see. They don't really document how to import for newer front ends.

Let's try
- Go to https://github.com/HabitRPG/habitica/blob/develop/website/client/main.js
- `npm i smartbanner.js`
- Then add `import smartbanner.js from 'smartbanner.js'; // eslint-disable-line
- Then maybe try importing the css like so `@import '~smartbanner.js/dist/smartbanner.min.css';` here https://github.com/HabitRPG/habitica/blob/develop/website/client/app.vue#L105

I'm not exactly sure though. Maybe @paglias has some idea.

And sorry for the odd workaround, but we'd like to keep the dependency separate.

flhat commented:
@TheHollidayInn No problem, thanks for the suggestion.  I'll play around with it and see what I can do.

paglias commented:
the steps suggested by @TheHollidayInn should work!

flhat committed:
Installed smartbanner.js as a module and imported css through app.vue

flhat commented:
Hmm, well, I got it worked out, but not quite sure how to handle the coverage issue...  :/

Alys commented:
Coveralls does that sometimes. You don't need to worry about it. :)

flhat commented:
@Alys Oh, okay.  So this pull request is ready for review then?

Alys commented:
@BlkStormy Good point, I've added the "ready for review" label.

paglias commented:
So this looks good but I think the icons used should be the Habitica apps icons instead of the google play / app store ones? In which case you can find them in website/static/presskit/Logo

paglias commented on website/client/main.js:
```diff
@@ -9,6 +9,7 @@ import getStore from './store';
 import StoreModule from './libs/store';
 import './filters/registerGlobals';
 import i18n from './libs/i18n';
+import smartbanner from 'smartbanner.js/dist/smartbanner'; // eslint-disable-line
```
I think the `// eslint-disable-line` can be removed

TheHollidayInn replied:
I think that we will get an unused var warning, but definitely double check me.

paglias replied:
You're right... I think then we should just do `import smartbanner.js/dist/smartbanner`

flhat committed:
Changed the logos to use the ones in the existing presskit directory and fixed the import line for the smartbanner component

flhat commented:
@paglias @TheHollidayInn thanks for reviewing!

1) Good call on the icon - updated and remove the 2 new images I had added

2) I changed the import to `import 'smartbanner.js/dist/smartbanner';`

paglias commented on website/client/app.vue:
```diff
@@ -103,7 +103,8 @@ div
 <style lang='scss'>
   @import '~client/assets/scss/colors.scss';
-
+  @import '~smartbanner.js/dist/smartbanner.min.css';
```
This should be loaded like other deps here https://github.com/HabitRPG/habitica/blob/develop/website/client/app.vue#L634 so in this case
`<style src="smartbanner.js/dist/smartbanner.min.css"></style>`

flhat replied:
Got it - I have this ready - thanks!

paglias commented on website/client/main.js:
```diff
@@ -9,6 +9,7 @@ import getStore from './store';
 import StoreModule from './libs/store';
 import './filters/registerGlobals';
 import i18n from './libs/i18n';
+import 'smartbanner.js/dist/smartbanner';
```
instead of loading the version in /dist which is already compiled using browserify we should load the original version that will then be compiled directly in Habitica, so the file to import should be `smartbanner.js/src/index.js`

flhat replied:
Hmm, the /src/ directory isn't included in the node module only /dist/ which contains smartbanner.js and smartbanner.min.js

paglias replied:
I see... In that case the non .min version will work

TheHollidayInn replied:
Shouldn't we be able to just import smartbanner.js? It is a module

paglias replied:
Looks like it's mostly published to npm but doesn't follow many conventions and if I remember correctly it doesn't have a main entry in package.json so i don't think it'll work, but we can try

flhat replied:
Updated - just committed the change

paglias reviewed (changes requested):
Left two comments, then it should be good to go!

flhat committed:
Changed smartbanner import to a src include for css and updated js import

paglias commented:
So I just tested this change and it works well, but I noticed we also have this warning 
<img width="495" alt="schermata 2018-08-20 alle 15 09 38" src="https://user-images.githubusercontent.com/589911/44342391-1be1ea80-a48b-11e8-8b1e-de32e679a2fa.png">

which I think should be removed if we add this?

@Tressley @TheHollidayInn

paglias commented:
This is the code to remove

```
diff --git a/website/client/components/header/index.vue b/website/client/components/header/index.vue
index fb7e2965c..c17a24c2a 100644
--- a/website/client/components/header/index.vue
+++ b/website/client/components/header/index.vue
@@ -35,8 +35,6 @@ div
         span.small-text(v-html="$t('inviteFriendsParty')")
         br
         button.btn.btn-primary(@click='createOrInviteParty()') {{ user.party._id ? $t('inviteFriends') : $t('startAParty') }}
-  a.useMobileApp(v-if="isAndroidMobile()", v-once, href="https://play.google.com/store/apps/details?id=com.habitrpg.android.habitica") {{ $t('useMobileApps') }}
-  a.useMobileApp(v-if="isIOSMobile()", v-once, href="https://itunes.apple.com/us/app/habitica-gamified-task-manager/id994882113?mt=8") {{ $t('useMobileApps') }}
 </template>

 <style lang="scss" scoped>
@@ -155,12 +153,6 @@ export default {
     ...mapActions({
       getPartyMembers: 'party:getMembers',
     }),
-    isAndroidMobile () {
-      return navigator.userAgent.match(/Android/i);
-    },
-    isIOSMobile () {
-      return navigator.userAgent.match(/iPhone|iPad|iPod/i);
-    },
     expandMember (memberId) {
       if (this.expandedMember === memberId) {
         this.expandedMember = null;
(END)
```

Tressley commented:
@paglias / @TheHollidayInn  -- I would like to keep the messaging for now. The smart banner doesn't contain any messaging around the lack of true mobile optimization.

I would, however, like to update the messaging with our visual styles. I believe the current implementation is a carry over from Keith's initial pass to get some mobile features available.

paglias commented:
Sounds good, in that case we can merge this PR

SabreCat commented:
Thanks for this, @BlkStormy, and welcome to the ranks of the Blacksmiths at tier 1! This PR will undergo some additional testing in a staging environment, and if all is well go live late next week.

Note that further contributor tiers require increasing amounts of effort from one to the next, so tier 2 will require multiple PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!
